examples/pytorch/__temporary__/dglnew/graph/graph.py

```diff
@@ -0,0 +1,181 @@
+class GraphStorage(object):
+    @property
+    def ntypes(self):
+        """The list of node types."""
+        pass
+
+    @property
+    def ndata(self):
+        """Node data.
+
+        For graphs with one node type it's a dict whose keys are feature names and values are
+        either tensors or FeatureStorage objects.
+
+        For multiple node types it's a dict of dict.  The outer keys are feature names
+        and the inner keys are node type names:
+
+        .. code::
+
+           self.ndata[feature_name][ntype]
+        """
+        pass
+
+    # Required in Link Prediction
+    @property
+    def etypes(self):
+        """The list of edge types."""
+        pass
+
+    # Required in Link Prediction
+    @property
+    def canonical_etypes(self):
+        """The list of canonical edge types (triplets of "source node type", "edge type",
+        and "destination node type")."""
+        pass
+
+    @property
+    def edata(self):
+        """Edge data.
+
+        For graphs with one edge type it's a dict whose keys are feature names and values are
+        either tensors or FeatureStorage objects.
+
+        For multiple edge types it's a dict of dict.  The outer keys are feature names
+        and the inner keys are canonical edge types or edge type names:
+
+        .. code::
+
+           self.edata[feature_name][etype]
+
+        Note that to maximize the compatibility with other DGL code it's better to support
+        both canonical edge types and edge type names as inner keys since users might use both.
+        """
+        pass
```

Comment:
Interface-wise I think `ndata` and `edata` should be stayed as-is rather than adding a new interface.  The graph storage's implementation is free to support writing to edata and ndata, etc., and should report error there if it's not supported.

Reply:
My big concern is implementing this correctly and efficiently is quite challenging.

Not only is there the combination of homogeneous graph vs. heterogeneous graph, etype vs. canonical etype, but then should each inner dictionary only contain the etypes and canonical etypes that have that feature, or should they have empty/None tensors associated with them? Also, the DGLHeteroGraph currently returns special view objects rather than dictionaries, which have hooks to allow modifying the underlying graph data, but for storage that return newly created dictionaries, such modifications would silently have no effect.

In terms of performance, creating dictionaries and new objects is quite expensive in python (in the context of mini-batch generation), this interface requires two-layers of objects just to get to the point of slicing a feature tensor.
However if we use a single function, not only will I not need to worry about the differences between homogeneous graphs and heterogeneous graphs, I can immediately drop into C code for efficiency:
```
def get_edge_features(feature_name, eids, output_device, edge_type):
  return _CAPI_MyCustomFunction(feature_name, eids, output_device, edge_type)
```
Which if our dataloader goes through each feature on each edge type, will be significantly faster than instantiating two levels of dictionaries / dictionary-like objects for each access.

Reply:
I think the major suggestion is to restrict the interface a little bit, e.g., disallow `g_store.ndata[feat_name]` which returns a dictionary of feature storage for each node/edge type. I like the idea but will tweak the proposal a little bit. How about:
```python
def get_ndata(name : str, ntype : str):
    # Get the feature storage associated with the given name and node type.
    pass

def get_edata(name : str, etype : Tuple[str, str, str]):
    # Get the feature storage associated with the given name and edge type.
    # Edge type is a string triplet.
    pass
```
Slicing will typically happen subsequently on the returned feature storage. See the `fetch` interface in the RFC.